presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PartialAggregationPushDown.java

```diff
@@ -91,6 +93,15 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
                     .map(functionRegistry::getAggregateFunctionImplementation)
                     .allMatch(InternalAggregationFunction::isDecomposable);
 
+            boolean partialRequired = node.getStep().equals(SINGLE) &&
```

Comment:
I couldn't extract too much logic out of it, but I have extracted and named empty/non empty checks.